[nnc] Add context-sensitive simplification for div/mod

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62138
* #60693
* __->__ #60688

Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

huiguoo committed:
[nnc] Add context-sensitive simplification for div/mod

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+// 1) (i + 24) / 6, i is the index var of a for-stmt and the range of i is
+// contained in [0, 6)
+//    -> 4
+// 2) (i + j*6) / 6, i, j are index vars of for-stmts and the range of i is
```
What if one of the constants is negative? Can you add a test for that?

navahgar reviewed (commented):
This looks promising. But simplifier changes are very tricky and can lead to silent bugs which would be very hard to debug. So, it would be good to add more tests, covering all possible boundary conditions here.

For the same reason, I would prefer at least another pair of eyes on this PR.

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+    const Expr* check_start =
+        IRSimplifier::simplify(new CompareSelect(start, new IntImm(0), kGE));
+    const Expr* check_end =
+        IRSimplifier::simplify(new CompareSelect(end, rhsScalar, kLE));
```
`end` comes from the `stop` of a `for-stmt`: the variable would reach its value so it should be `kLE`.

navahgar replied:
Are you saying `stop` refers to the last iteration index value? 

IIUC it is one more than the index value in the last iteration. See how it is used in llvm codegen here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/llvm_codegen.cpp#L1327

Am I missing something?

huiguoo replied:
Yeah, that's what I tried to say: `end` is one more than the index value, i.e., the index variable would NOT reach the `stop` value (i.e., `end`). To make sure that the index value is less than `rhsScalar`, we only need to check whether `end` is equal to or less than `rhsScalar`.

Just realized there is a typo in my previous comment; replaced `would reach` with `would not reach`. Sorry for the misunderstanding :smiley:

navahgar replied:
Can you add this as a comment here?

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+    const Expr* check_start =
+        IRSimplifier::simplify(new CompareSelect(start, new IntImm(0), kGE));
+    const Expr* check_end =
+        IRSimplifier::simplify(new CompareSelect(end, rhsScalar, kLE));
```
I think it should be `kLE`

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+      const Expr* ret = IRSimplifier::simplify(new Div(main, rhsScalar));
+
+      // simplify type 1 exprs: '(i+24)/6'
+      const Expr* mod_check = IRSimplifier::simplify(new Mod(main, rhsScalar));
```
This simplifies `(i + 24) / 6` into `4` where `i` is a non-negative int less than 6 and 24 is divisible by 6. The mod check verifies that 24 is divisible by 6. If mod check failed, we can not remove `i` in the numerator.

navahgar replied:
Please add this as a comment as well.

huiguoo replied:
That's true. In theory, if `var%rhsScalar + main%rhsScalar < rhsScalar` then it is valid to distribute the `mod` operator: `(var + main)%rhsScalar --> var%rhsScalar + main%rhsScalar`.  In this case `(i + 25)%6 where i<5` can be simplified into `i + 1`.

But we may need to rethink the two-step approach (distribution + range based simplification) because the generalized check also requires the range info. of var and it can immediately reach the final result: `i+1` without step 2.

navahgar reviewed (commented):
I don't have any other major comments. I ll let another person review this and approve.

One suggestion I have is to add some debug info while simplifying using this logic, so that, if we want to debug something here, just setting the debug flag would be enough to see what is going on here. I don't think the rest of `ir_simplifier` have such debug info, but we should probably add that as well in future.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+}
+
+// Simplify mod using distributive laws for the following cases:
+// 1) (i + 24) % 6, i is the index var of a for-stmt and the range of i is
```
I wonder if the implementation would be simpler if we broke the optimization into two steps:
1) `(x + y) % CONST` --> `x % CONST + y % CONST`
2) `x % CONST -> x` if `0 <= x < CONST`

Similarly for div it would be:
1) `(x + y) / CONST` --> `x / CONST + y / CONST` if `x % CONST == 0` or `y % CONST == 0`
2) `x / CONST -> 0` if `0 <= x < CONST`

Here only the second step would be using range info. One downside of this approach is that we could distribute divs/mods without the subsequent optimization, but I wonder what you think?

ZolotukhinM replied:
Btw, #52580 is related to this. Maybe we should fix it too while we're at it.

huiguoo replied:
I like this idea! I already have a PR for #52580: PR #60693 but it is not a clean solution to me. I was wondering how to combine the two more neatly and your suggestion seems to be the right way to go. 

Basically, in the first step we distribute divs/mods if `x % CONST == 0 or y % CONST == 0` and perform and the range based simplification in second step. The first step is also a simplification even without the subsequent optimization: 
`(i + j*6)%6 --> i%6` or `(i + j*6)/6 --> i/6 + j`

After the first step, the mod expr is still a mod which should not cause any trouble. The distribution of div is also a good thing? It simplified the expression into a polynomial-like expression for subsequent simplifications on polynomials?

ZolotukhinM replied:
Discussed this offline and agreed that having a simplification with the two steps combined along with a separate simplification for (2) is a way to go, because (1) needs a range info anyway.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+        if (got == var_bound_info.end()) {
+          return nullptr;
+        }
+        check_start = IRSimplifier::simplify(
+            new CompareSelect(start, new IntImm(0), kGE));
+        if (check_start->isConstant() && immediateEquals(check_start, 1)) {
+          return var_key;
+        }
```
We've already checked that on line 2609, right? What am I missing here?

huiguoo replied:
This is different from the check on line 2609. I am checking the start value of another var. See comments above.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+      }
+
+      // simplify type 2 exprs: '(i+j*6)%6'
+      auto ret_var = dynamic_cast<const Var*>(ret);
+      if (ret_var) {
```
I think this is legal not only when `ret` is `Var`, but generally when `ret` is not `Div`.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+      auto ret_var = dynamic_cast<const Var*>(ret);
+      if (ret_var) {
+        auto got = var_bound_info.find(ret_var);
+        if (got == var_bound_info.end()) {
+          return nullptr;
+        }
```
This doesn't seem to be used anywhere, why do we need it?

huiguoo replied:
Bug/typo: it should be used to initialize `start` used in the following check `check_start`.

ZolotukhinM commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -1235,6 +1235,118 @@ TEST(Simplify, SimplifyDiv) {
+  KernelScope kernel_scope;
+  // Stmt to simplify:
+  // for (int i = 0; i < 5; i++) {
+  //  for (int j = 0; j < 5; j++) {
+  //    A[i, j] = (i + 6*j) / 6;
```
The optimization should be valid if j iterates beyond 6, right? Could you please add a test for this?

ZolotukhinM replied:
This comment is still relevant.

huiguoo replied:
Yeah, it should be valid if j is greater than 6. Will add a test!

ZolotukhinM commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -1235,6 +1235,118 @@ TEST(Simplify, SimplifyDiv) {
+  // Stmt to simplify:
+  // for (int i = 0; i < 5; i++) {
+  //  for (int j = 0; j < 5; j++) {
+  //    A[i, j] = (i + 6*j) % 6;
```
Ditto.

ZolotukhinM reviewed (commented):
Overall looks good to me, please find some comments below!

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+  if (!lhs || !rhs) {
+    return nullptr;
+  }
+  auto* lhsAdd = dynamic_cast<const Add*>(lhs);
+  const Expr* rhsScalar = rhs->isConstant() ? rhs : nullptr;
```
The polynomial representation somehow makes it harder to decode the pattern. I have tried it but found that I am not able to easily retrieve the `var`：my inquire to the range info of the `var` always returns false. I didn't spend too much time on it but the the polynomial form does not seem to be a necessity for this.

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+  }
+
+  if (var_key != nullptr) {
+    auto got = var_bound_info.find(var_key);
```
You mean its type/content? std::pair<const Expr*, const Expr*>: start/end value of its range.

ZolotukhinM replied:
I mean I don't understand what the name "got" means here. Is it some abbreviation?

huiguoo replied:
the past tense of 'get', lol

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2490,6 +2490,185 @@ Stmt* SimplifierUnderContext::mutate(const For* v) {
+        }
+        check_start = IRSimplifier::simplify(
+            new CompareSelect(start, new IntImm(0), kGE));
+        if (check_start->isConstant() && immediateEquals(check_start, 1)) {
```
This optimizes `(i+j*6)/6` whether `0<=i<6` and `j>=0`, so `(i+j*6)/6-->j`. 
The `check_start` checks the start value of j and ensures that it is not negative. We not care the end value of `j`'s range.

huiguoo replied:
I realized that the `start` is not initialized with `got->second.first`. Will add it!

ZolotukhinM replied:
Why do we need `j>=0`? I think if `0<=i<6` then `(i+j*6)/6` is always equal to `j`, no matter what `j` is (for the mod case it's different and there we do need to check if it's not negative).

huiguoo replied:
It depends on the compiler? I tested negative numbers for `j` on my laptop: both gcc and python returns a different number from `j`. e.g., if `i=3` and `j=-1`, `(i+j*6)/6` equals to 0.

huiguoo replied:
Same for NNC: if `i=3` and `j=-1`, (i+j*6)/6 equals to 0

So we cannot simplify `(i+j*6)/6` to `j` if `j` is negative at this moment.  Would have to wait till pytorch and NNC correct the semantics of integer division.

ZolotukhinM commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -1235,6 +1235,118 @@ TEST(Simplify, SimplifyDiv) {
+  VarHandle i("i", kInt);
+  BufHandle a_buf("A", {5}, kInt);
+  auto forstmt = For::make(
+      i, 0, 5, Store::make(a_buf, {i}, (i + ExprHandle(24)) / ExprHandle(6)));
```
Nit: I think you can write just "`(i + 24) / 6`", there is no need in wrapping int immediates with `ExprHandle` there. This applies to other tests too.

ZolotukhinM commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -1235,6 +1235,118 @@ TEST(Simplify, SimplifyDiv) {
+  auto forstmt = For::make(
+      i, 0, 5, Store::make(a_buf, {i}, (i + ExprHandle(24)) / ExprHandle(6)));
+
+  const Stmt* simplified = IRSimplifier::simplify(forstmt);
```
Nit: for_stmt?

ZolotukhinM reviewed (approved):
Looks good to go once remarks are addressed!

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@navahgar I'll add the logging info in another PR.

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).

huiguoo committed:
Update on "[nnc] Add context-sensitive simplification for div/mod"


Differential Revision: [D29373313](https://our.internmc.facebook.com/intern/diff/D29373313)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373313).
